### DISCUSSION TITLE
Cannot find module "/react/jsx-runtime"

### DISCUSSION THREAD

**Discussion by cristianpoleyJS** [vitest-dev/vitest#2055] [Q&A]
I'm migrating a project from `jest` to `vitest`, but I'm receiving this error when try to run a simple test:

```bash
FAIL  src/__tests__/components/ButtonsSwitch.unit.test.tsx [ src/__tests__/components/ButtonsSwitch.unit.test.tsx ]
Error: Cannot find module '/Users/user/projects/project/webapp/node_modules/react/jsx-runtime' imported from /Users/user/projects/project/webapp/node_modules/react-dnd/dist/core/DndProvider.js
Did you mean to import react/jsx-runtime.js?
```

This is my `vite.config.ts`:

```js
/// <reference types="vite/client" />
import { optimizeLodashImports } from '@optimize-lodash/rollup-plugin';
import react from '@vitejs/plugin-react';
import path from 'path';
import { visualizer } from 'rollup-plugin-visualizer';
import { defineConfig } from 'vite';
import envCompatible from 'vite-plugin-env-compatible';
import OptimizationPersist from 'vite-plugin-optimize-persist';
import PkgConfig from 'vite-plugin-package-config';
import svgrPlugin from 'vite-plugin-svgr';
import tsconfigPaths from 'vite-tsconfig-paths';

var config = defineConfig((config) => ({
  plugins: [
    react({
      fastRefresh: process.env.NODE_ENV !== 'test',
    }),
    tsconfigPaths(),
    PkgConfig(),
    OptimizationPersist(),
    svgrPlugin(),
    envCompatible({
      prefix: 'REACT_APP_',
    }),
    optimizeLodashImports(),
    visualizer({
      open: false,
      gzipSize: true,
    }),
  ],
  css: { preprocessorOptions: { css: { charset: false }, scss: { charset: false } } },
  resolve: {
    alias: [
      { find: /^@themes/, replacement: path.join(__dirname, 'src/themes') },
      {
        find: './runtimeConfig',
        replacement: './runtimeConfig.browser',
      },
      {
        find: 'react/jsx-runtime',
        replacement: 'react/jsx-runtime.js',
      },
      {
        find: '@',
        replacement: path.resolve(__dirname, 'src'),
      },
    ],
  },
  build: {
    chunkSizeWarningLimit: 6200, // Kb
    outDir: 'build',
    sourcemap: config.mode === 'test' ? false : 'hidden',
  },
  server: {
    port: 3000,
    proxy: {
      '/api': {
        changeOrigin: true,
        target: process.env.REACT_APP_BACKEND_URL,
        rewrite: (path) => path.replace(/^\/api/, ''),
      },
    },
  },
  test: {
    globals: {
      'ts-jest': {
        diagnostics: true,
        useESM: true,
      },
    },
    environment: 'jsdom',
    setupFiles: './src/setupTests.ts',
    // you might want to disable it, if you don't have tests that rely on CSS
    // since parsing CSS is slow
    css: true,
    dir: 'src',
    exclude: ['cypess/**.*, playwright/**.*'],
    deps: {
      inline: ['vitest-canvas-mock'],
    },
    mockReset: true,
    testTimeout: 15000,
    allowOnly: true,
    restoreMocks: true
  },
}));

export default config;
```

In my `setupTest` file, I'm mocking a store:

```vi.mock('./store', () => vi.importActual('./utils/testing/store'));```

In `utils/testing/store` I am creating a mock of the store (with `@reduxjs/toolkit`) of all modules. It fails with some modules and not with others.

I read this [Issue](https://github.com/vitest-dev/vitest/issues/1572) but I already have the react plugin on my `vite.config.js`. 
Any idea? Thanks a lot

**Comment by sheremet-va** [comment]
Aliases don't work in `node_modules` by default. You need to `deps.inline` it, or enable `deps.registerNodeLoader`.

**Reply by cristianpoleyJS** [reply]
Thank you for your quick response.

After changing `deps` to:

```js
// ...
    deps: {
      inline: true,
    },
// ...
```

I have this error (`TypeError: exports.hasOwnProperty is not a function`): 
```bash
 FAIL  src/domains/__tests__/components/ButtonsSwitch.unit.test.tsx [ src/domains/__tests__/components/ButtonsSwitch.unit.test.tsx ]
 ❯ Object.__exportStar node_modules/tslib/tslib.js:153:58
TypeError: exports.hasOwnProperty is not a function
 ❯ node_modules/@sentry/react/src/index.ts:1:0
```

Specifications:
```
node --version -> v16.13.0
npm --version -> 8.1.0
```

Otherwise, if I use `deps.registerNodeLoader` I have this error:

```bash
FAIL  src/domains/__tests__/components/ButtonsSwitch.unit.test.tsx [ src/domains/__tests__/components/ButtonsSwitch.unit.test.tsx ]
Error: Vitest was initialized with native Node instead of Vite Node.

One of the following is possible:
- "vitest" is imported outside of your tests (in that case, use "vitest/node" or import.meta.vitest)
- "vitest" is imported inside "globalSetup" (use "setupFiles", because "globalSetup" runs in a different context)
- Your dependency inside "node_modules" imports "vitest" directly (in that case, inline that dependency, using "deps.inline" config)
- Otherwise, it might be a Vitest bug. Please report it to https://github.com/vitest-dev/vitest/issues
```

**Reply by sheremet-va** [reply]
> After changing `deps` to:

You can just add `/react-dnd/` to it.

**Reply by cristianpoleyJS** [reply]
I just added it but I have the same error.
```js
//...
    deps: {
      inline: ['vitest-canvas-mock', '/react-dnd/'],
    },
//...
```

**Reply by sheremet-va** [reply]
```js
//...
    deps: {
      inline: ['vitest-canvas-mock', /react-dnd/],
    },
//...
```

**Reply by cristianpoleyJS** [reply]
Oh right, it can be regexp, but with this change the execution is frozen:

<img width="471" alt="image" src="https://user-images.githubusercontent.com/15108191/190924515-e28494d9-6fc8-4987-a57c-6c29298b350d.png">

**Reply by cristianpoleyJS** [reply]
@sheremet-va 
I update with what I have found. Depending on the configuration different things happen to me:

1. With this configuration, when I launch `npm run test` it only runs a few tests, but there are many others within the same path that are not launched.
If I launch `npm run test <dir_file>` the execution freezes like [here](https://github.com/vitest-dev/vitest/discussions/2055#discussioncomment-3675805).
```javascript
//...
    deps: {
      inline: ['vitest-canvas-mock', /react-dnd/],
    },
//...
```

2. When I launch `npm run test` it launches all the tests I have but when I launch the `npm run test <dir_file>` the error I get is this again (`Error: Cannot find module '/Users/user/projects/project/webapp/node_modules/react/jsx-runtime' imported from /Users/user/projects/project/webapp/node_modules/react-dnd/dist/core/DndProvider.js
Did you mean to import react/jsx-runtime.js?`).
```javascript
//...
    deps: {
      inline: ['vitest-canvas-mock'],
      external: [/react-dnd/]
    },
//...
```

3. When I launch `npm run test` or `npm run test <dir_file>`, I have the error `Vitest was initialized with native Node instead of Vite Node.`
```javascript
//...
    deps: {
      registerNodeLoader: true,
    },
//...
```

I don't know if this is a bug or not, or maybe I'm missing something from the configuration I shared at the origin of the post.

**Reply by LucasMonteiroDito** [reply]
Hey, i just added in this way and works for me 
![image](https://github.com/vitest-dev/vitest/assets/119513757/392d5c75-74d6-45ac-b08e-8e1e051dac3e)

**Reply by ChintanAcharya** [reply]
I set the `server.deps.fallbackCJS` option to `true` (https://vitest.dev/config/#server-deps-fallbackcjs). That worked too.
